rustls/src/server/hs.rs

```diff
@@ -213,22 +212,25 @@ impl ExtensionProcessing {
         config: &ServerConfig,
         cx: &mut ServerContext<'_>,
     ) -> Result<(), Error> {
-        let requires_server_rpk = config
+        let server_supports = if config
             .cert_resolver
-            .only_raw_public_keys();
-        let client_allows_rpk = hello
-            .server_certificate_extension()
-            .map(|certificate_types| certificate_types.contains(&CertificateType::RawPublicKey))
-            .unwrap_or(false);
-
-        let raw_key_negotation_params = RawKeyNegotiationParams {
-            peer_supports_raw_key: client_allows_rpk,
-            local_expects_raw_key: requires_server_rpk,
-            extension_type: ExtensionType::ServerCertificateType,
+            .only_raw_public_keys()
+        {
+            CertificateType::RawPublicKey
+        } else {
+            CertificateType::X509
         };
+        let client_supports = hello
+            .server_certificate_extension()
+            .map(|certificate_types| certificate_types.to_vec())
+            .unwrap_or_default();
 
         self.process_cert_type_extension(
-            raw_key_negotation_params.validate_raw_key_negotiation(),
+            check_certificate_type(
```

Comment:
What's the reason to keep `process_cert_type_extension()` and `check_certificate_type()` split? It looks like we could save some duplication/overhead by merging them.